### PR TITLE
[WIP] Basic support for itunes rss tags

### PR DESCRIPTION
**Pull request by anderbubble**
I'm working on enhancing Nikola to support publishing a podcast RSS feed. Most of this work is going into a "postcast" plugin I'm working on, and will submit or publish eventually; but some of what I'm trying to do requires support in Nikola itself.

This PR aims to support [iTunes-specific tags](https://help.apple.com/itc/podcasts_connect/#/itcb54353390) in RSS feeds.

Other related PRs I'm working on:

* #2676 (copyright tag in rss feeds)
* #2675 (explicit / static guids)

I'd also be interested in direction for where and how to publish a plugin, when that's ready.

## todo

* remove references to postcast

### CODE REVIEW

File: nikola/utils.py [hunk 1]
```diff
@@ -1272,13 +1286,66 @@ def publish_extensions(self, handler):
             })
             handler.endElement("atom:link")
 
+        if self.itunes_author:
```
Comment (Kwpolska) [inline nikola/utils.py]: You could duplicate less code, eg. with `getattr` or some iTunes data dictionary.

**Comment by Kwpolska** [comment]
Looks good (apart from a lot of duplicated code). Please update CHANGES.txt, AUTHORS.txt, and the manual.

File: nikola/utils.py [hunk 2]
```diff
@@ -1255,6 +1255,20 @@ class ExtendedRSS2(rss.RSS2):
...
+                 itunes_explicit=None, **kwargs):
+        self.itunes_author = itunes_author
+        self.itunes_summary = itunes_summary
+        self.itunes_name = itunes_name
+        self.itunes_email = itunes_email
+        self.itunes_image = itunes_image
+        self.itunes_categories = itunes_categories
+        self.itunes_explicit = itunes_explicit
+
+        return rss.RSS2.__init__(self, **kwargs)
```
Comment (Kwpolska) [inline nikola/utils.py]: Don’t use `return` here, instead call `rss.RSS2.__init__` at the start of this function. (old-style class in Python 2, so `super()` does not work(

**Comment by anderbubble** [comment]
First, thank you for reviewing this pull request!

I returned `__init__` from `__init__` because the code for `Item` already did.[^1] Should I clean that up at the same time?

As for the duplicated code: can you provide a pointer to somewhere else in the codebase that does something similar better? (It'd just be nice to have an example of what idiom is considered most normal in this codebase.)

I'm also not happy with my handling of the `Image` url. Any suggestions for how to do it better?

[^1]: https://github.com/getnikola/nikola/pull/2674/files#diff-cca99a67a7b16498786d42c7f2f89444L1283

**Comment by Kwpolska** [comment]
1. Yes, you can do that as well.
2. I don’t think we have anything like this. Here’s one example:

```py
itunes = {'itunes:subtitle': 'foo'}
# in publish_extensions
for name in ['itunes:author', 'itunes:subtitle', 'itunes:summary', 'itunes:duration']:
    if name in itunes:
        handler.startElement(name, {})
        handler.characters(itunes[name])
        handler.endElement(name)
# special handling for image and explicit stays
```

3. You could ask `url_replacer` with `url_type='absolute'`.

File: nikola/nikola.py [hunk 3]
```diff
@@ -1649,7 +1649,7 @@ def apply_shortcodes(self, data, filename=None, lang=None, with_dependencies=Fal
 
     def generic_rss_renderer(self, lang, title, link, description, timeline, output_path,
                              rss_teasers, rss_plain, feed_length=10, feed_url=None,
-                             enclosure=_enclosure, rss_links_append_query=None):
+                             enclosure=_enclosure, rss_links_append_query=None, itunes=None):
```
Comment (felixfontein) [inline nikola/nikola.py]: Souldn't the default value for `itunes` be `False` and not `None`? After all, it is only used like a boolean.

**Comment by anderbubble** [comment]
It is my habit to use `None` for a default value so that the underlying
code can distinguish between "unspecified" and "explicitly false."

But f that's counter to expectation here, I'm fine to change it.

**Comment by Kwpolska** [comment]
`None` is good in this case IMO.

**Comment by anderbubble** [comment]
I see now that I've added explicit references to my "postcast" plugin in this Nikola core PR; so I'm going to have to factor that out.

File: nikola/nikola.py [hunk 4]
```diff
@@ -1711,6 +1711,17 @@ def generic_rss_renderer(self, lang, title, link, description, timeline, output_
...
+            if itunes:
+                for tag in ('subtitle', 'duration', 'explicit'):
+                    key = 'itunes_{}'.format(tag)
+                    args[key] = post.meta[lang].get(key)
+                args['itunes_author'] = (
+                    post.meta[lang].get('itunes_author')
+                    or post.author(lang))
+                args['itunes_summary'] = post.meta[lang].get('itunes_summary') or data
+                if 'itunes_image' in post.meta[lang]:
+                    args['itunes_image'] = self.url_replacer(post.permalink(), post.meta[lang]['itunes_image'], lang, 'absolute')
```
Comment (Kwpolska) [inline nikola/nikola.py]: Just keep in mind that the permalink won’t affect parsing of `itunes_image` paths, hey should be relative to site root (in your posts).

File: nikola/nikola.py [hunk 5]
```diff
@@ -1711,6 +1711,17 @@ def generic_rss_renderer(self, lang, title, link, description, timeline, output_
             if post.author(lang):
                 rss_obj.rss_attrs["xmlns:dc"] = "http://purl.org/dc/elements/1.1/"
 
+            if itunes:
+                for tag in ('subtitle', 'duration', 'explicit'):
+                    key = 'itunes_{}'.format(tag)
+                    args[key] = post.meta[lang].get(key)
+                args['itunes_author'] = (
```
Comment (Kwpolska) [inline nikola/nikola.py]: Write that on one line, will look better and fix a style checker complaint.

File: nikola/utils.py [hunk 6]
```diff
@@ -1255,6 +1255,25 @@ class ExtendedRSS2(rss.RSS2):
...
+        self.itunes_name = itunes_name
+        self.itunes_email = itunes_email
+        self.itunes_image = itunes_image
+        self.itunes_categories = itunes_categories
+        self.itunes_explicit = itunes_explicit
+
+        # It's an old style class
+        rss.RSS2.__init__(self, **kwargs)
+
+    def _itunes_attributes (self):
```
Comment (Kwpolska) [inline nikola/utils.py]: Remove the space before `(self)` here.

File: nikola/utils.py [hunk 7]
```diff
@@ -1272,15 +1291,49 @@ def publish_extensions(self, handler):
...
+                    handler.startElement("itunes:category", {'text': subcategory})
+                    handler.endElement("itunes:category")
+                handler.endElement("itunes:category")
+
 
 class ExtendedItem(rss.RSSItem):
     """Extended RSS item."""
 
-    def __init__(self, **kw):
+    def __init__(self, itunes_author=None, itunes_subtitle=None,
```
Comment (Kwpolska) [inline nikola/utils.py]: It might be better to use the `kw` dict as well for your itunes attributes.

File: nikola/nikola.py (hunk 4 above)
Comment (anderbubble) [inline nikola/nikola.py]: That hasn't been my experience. With this code, the image is located relative to the post's location, and then an absolute url to that location is generated.

File: nikola/utils.py (hunk 6 above)
Comment (anderbubble) [inline nikola/utils.py]: Oops. That's a minor stylistic affectation I use in my own code, but don't intend to inflict on anyone. Just a habit. :)

File: nikola/utils.py (hunk 7 above)
Comment (anderbubble) [inline nikola/utils.py]: Doesn't that kind of reading-into-kwargs defeat static code analysis? If you don't declare what arguments you take in your function definition, you're unnecessarily hiding information.

At least, that's been my take on it in the past.

Comment (Kwpolska) [inline nikola/utils.py]: I would prefer to keep it this way if it was started this way — perhaps there’s a reason for using `kw` there?

**Comment by anderbubble** [comment]
I'm actually going a different way on this; leave this open for now, but I'd prefer to merge #2686 rather than this, and do my iTunes-specific stuff in my plugin rather than pollute core.

**Comment by felixfontein** [comment]
Since now #2686 is merged, can we close this @anderbubble?

**Comment by anderbubble** [comment]
@felixfontein for sure. Thanks.

**Pull request closed without merging**